File: src/main/java/ru/arc/stock/StockMarket.java
package ru.arc.stock;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;
import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.configs.ConfigManager;
import ru.arc.configs.StockConfig;
import ru.arc.repository.LegacyRedisRepo;

import static ru.arc.util.Logging.debug;
import static ru.arc.util.Logging.error;
import static ru.arc.util.Logging.info;

@RequiredArgsConstructor
public class StockMarket {
    private static BukkitTask updateTask, dividendTask;
    @Setter
    private static StockClient client;
    //private static Map<String, Stock> stockMap = new ConcurrentHashMap<>();

    private static final Map<String, ConfigStock> configStocks = new ConcurrentHashMap<>();
    private static final Config config = ConfigManager.of(ARC.getInstance().getDataPath(), "stocks/stock.yml");
    private static LegacyRedisRepo<Stock> repo;

    public static void init() {
        if(!config.bool("enabled", false)) {
            info("Stocks are disabled");
            return;
        }
        if (repo == null) {
            repo = LegacyRedisRepo.builder(Stock.class)
                    .loadAll(true)
                    .redisManager(ARC.redisManager)
                    .storageKey("arc.stocks")
                    .updateChannel("arc.stocks_update")
                    .id("stocks")
                    .backupFolder(ARC.getInstance().getDataFolder().toPath().resolve("backups/stocks"))
                    .saveInterval(20L)
                    .build();
        }
        startTasks();
        HistoryManager.init();
    }

    public static void startTasks() {
        cancelTasks();

        updateTask = new BukkitRunnable() {
            @Override
            public void run() {
                if (!StockConfig.mainServer) return;
                //// Logging removed - was using @Slf4j
                Map<String, Double> updates = new HashMap<>();

                boolean fetchedCrypto = false;
                for (var entry : configStocks.entrySet()) {
                    try {
                        Stock current = repo.getNow(entry.getKey());
                        long lastUpdated = current == null ? 0 : current.getLastUpdated();
                        if (System.currentTimeMillis() - lastUpdated > StockConfig.stockRefreshRate * 1000L) {
                            if (entry.getValue().type == Stock.Type.CRYPTO) {
                                if (fetchedCrypto) continue;
                                updates.putAll(client.cryptoPrices());
                                fetchedCrypto = true;
                                continue;
                            }
                            updates.put(entry.getKey(), client.price(entry.getValue()));
                        }
                    } catch (Exception e) {
                        error("Error fetching data for: {}", entry.getKey(), e);
                    }
                }

                // Logging removed - was using @Slf4j

                for (var entry : updates.entrySet()) {
                    try {
                        String symbol = entry.getKey().toUpperCase();
                        double price = entry.getValue();
                        double finalPrice = price;
                        Stock current = repo
                                .getOrCreate(symbol, () -> configStocks.get(symbol)
                                        .toStock(finalPrice, 0, System.currentTimeMillis(), 0))
                                .get();


                        if (price < 0 || price > 1_000_000) {
                            if (current.getPrice() < 0 || current.getPrice() > 1_000_000) {
                                error("Price for " + symbol + " is invalid: " + price);
                                continue;
                            }
                            price = current.getPrice();
                        }
                        HistoryManager.add(symbol, price);

                        current.setPrice(price);
                        current.setLastUpdated(System.currentTimeMillis());
                        if (current.getType() == Stock.Type.STOCK) {
                            current.setDividend(current.getPrice() * StockConfig.dividendPercentFromPrice);
                            if (current.getDividend() > 10_000) {
                                error("Dividend for " + symbol + " is invalid: " + current.getDividend());
                                current.setDividend(0);
                            }
                        }
                        current.setDirty(true);

                        // Logging removed - was using @Slf4j

                        StockPlayerManager.updateAllPositionsOf(symbol);
                    } catch (Exception e) {
                        error("Error updating stock: {}", entry.getKey(), e);
                    }

                }
            }
        }.runTaskTimerAsynchronously(ARC.getInstance(), 20L, 10 * 20L);

        dividendTask = new BukkitRunnable() {
            @Override
            public void run() {
                if (!StockConfig.mainServer) return;
                stocks().stream()
                        .filter(stock -> stock.getDividend() > 0.000001)
                        .filter(s -> System.currentTimeMillis() - s.getLastTimeDividend() >= 23 * 60 * 60 * 1000L)
                        .peek(stock -> stock.setLastTimeDividend(System.currentTimeMillis()))
                        .peek(stock -> stock.setDirty(true))
                        .map(Stock::getSymbol)
                        .forEach(StockPlayerManager::giveDividend);
            }
        }.runTaskTimer(ARC.getInstance(), 100L, 20L * 60);
    }

    public static void saveHistory() {
        if(!config.bool("enabled", false)) {
            info("Stocks are disabled");
            return;
        }
        HistoryManager.saveHistory();
    }

    public static void cancelTasks() {
        if(!config.bool("enabled", false)) {
            info("Stocks are disabled");
            return;
        }
        if (updateTask != null && updateTask.isCancelled()) updateTask.cancel();
        if (dividendTask != null && dividendTask.isCancelled()) dividendTask.cancel();
        HistoryManager.cancelTasks();
    }


    public static Stock stock(String symbol) {
        return repo.getNow(symbol);
    }

    public static Collection<Stock> stocks() {
        return repo.all();
    }

    @SuppressWarnings("unchecked")
    public static void loadStockFromMap(Map<?, ?> map) {
        try {
            ConfigStock stock = ConfigStock.deserialize((Map<String, Object>) map);
            stock.setSymbol(stock.getSymbol().toUpperCase());
            configStocks.put(stock.getSymbol(), stock);

            if (repo == null) return;

            var current = repo.getNow(stock.getSymbol());
            if (current == null) return;
            current.setLore(stock.getLore());
            current.setDisplay(stock.getDisplay());
            current.setIcon(stock.getIcon());
            current.setMaxLeverage(stock.getMaxLeverage());
            current.setType(stock.getType());
            current.setDirty(true);
        } catch (Exception e) {
            error("Error loading stock from map: {}", map, e);
            debug("[stock] failed to deserialize stock entry keys={}", map != null ? map.keySet() : null);
        }
    }

    public static boolean isEnabledStock(Stock stock) {
        if (stock == null) return false;
        return configStocks().stream().map(ConfigStock::getSymbol).collect(Collectors.toSet()).contains(stock.getSymbol());
    }


    public static Collection<ConfigStock> configStocks() {
        return configStocks.values();
    }
}

File: src/main/kotlin/ru/arc/stock/StockMarket.kt
package ru.arc.stock

import ru.arc.config.StockConfig
import ru.arc.repository.CachedRepository
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info
import java.util.concurrent.ConcurrentHashMap

object StockMarket {

    private val configStocks = ConcurrentHashMap<String, ConfigStock>()
    lateinit var stockRepo: CachedRepository<Stock>
    @Volatile
    private var client: StockClient? = null

    @JvmStatic fun stock(symbol: String): Stock? = stockRepo.getNow(symbol)

    @JvmStatic fun stocks(): Collection<Stock> = stockRepo.allNow()

    @JvmStatic fun configStocks(): Collection<ConfigStock> = configStocks.values

    @JvmStatic
    fun isEnabledStock(stock: Stock?): Boolean {
        if (stock == null) return false
        return configStocks.containsKey(stock.symbol)
    }

    @JvmStatic
    fun setClient(stockClient: StockClient) {
        val previous =
            synchronized(this) {
                client.also { client = stockClient }
            }
        if (previous !== stockClient) previous?.close()
    }

    @JvmStatic
    fun closeClient() {
        val previous =
            synchronized(this) {
                client.also { client = null }
            }
        previous?.close()
    }

    internal fun resetConfiguration() {
        configStocks.clear()
        closeClient()
    }

    @JvmStatic
    fun loadStockFromMap(map: Map<*, *>) {
        try {
            @Suppress("UNCHECKED_CAST")
            val stock = ConfigStock.deserialize(map as Map<String, Any>)
            stock.symbol = stock.symbol.uppercase()
            configStocks[stock.symbol] = stock

            if (!::stockRepo.isInitialized) return
            val current = stockRepo.getNow(stock.symbol) ?: return
            current.lore = stock.lore
            current.display = stock.display
            current.icon = stock.icon
            current.maxLeverage = stock.maxLeverage
            current.type = stock.type
            stockRepo.markDirty(current)
        } catch (e: Exception) {
            error("Error loading stock from map: {}", map, e)
        }
    }

    suspend fun updateStocks() {
        if (!StockConfig.mainServer) return
        val activeClient = client ?: return

        val updates = mutableMapOf<String, Double>()
        var fetchedCrypto = false

        for ((symbol, configStock) in configStocks) {
            try {
                val current = stockRepo.getNow(symbol)
                val lastUpdated = current?.lastUpdated ?: 0L
                if (System.currentTimeMillis() - lastUpdated > StockConfig.stockRefreshRate * 1000L) {
                    if (configStock.type == Stock.Type.CRYPTO) {
                        if (fetchedCrypto) continue
                        updates.putAll(activeClient.cryptoPrices())
                        fetchedCrypto = true
                        continue
                    }
                    updates[symbol] = activeClient.price(configStock)
                }
            } catch (e: Exception) {
                error("Error fetching data for: {}", symbol, e)
            }
        }

        for ((symbol, rawPrice) in updates) {
            try {
                val upperSymbol = symbol.uppercase()
                val configStock = configStocks[upperSymbol] ?: continue
                val current = stockRepo.getOrCreate(upperSymbol) {
                    configStock.toStock(rawPrice, 0.0, System.currentTimeMillis(), 0)
                }.getOrNull() ?: continue

                var price = rawPrice
                if (price < 0 || price > 1_000_000) {
                    if (current.price < 0 || current.price > 1_000_000) {
                        error("Price for $upperSymbol is invalid: $price")
                        continue
                    }
                    price = current.price
                }

                HistoryManager.add(upperSymbol, price)
                current.price = price
                current.lastUpdated = System.currentTimeMillis()
                if (current.type == Stock.Type.STOCK) {
                    current.dividend = effectiveDividendPerShare(current)
                    if (current.dividend > 10_000) {
                        error("Dividend for $upperSymbol is invalid: ${current.dividend}")
                        current.dividend = 0.0
                    }
                }
                stockRepo.markDirty(current)

                StockPlayerManager.updateAllPositionsOf(upperSymbol)
            } catch (e: Exception) {
                error("Error updating stock: {}", symbol, e)
            }
        }
    }

    fun payDividends() {
        if (!StockConfig.mainServer) return
        val now = System.currentTimeMillis()
        stocks()
            .filter { effectiveDividendPerShare(it) > 0.000001 }
            .filter { isDividendDue(it, now) }
            .forEach { stock ->
                stock.lastTimeDividend = now
                stockRepo.markDirty(stock)
                StockPlayerManager.giveDividend(stock.symbol)
            }
    }

    internal fun effectiveDividendRate(): Double {
        val configured = StockConfig.dividendPercentFromPrice
        val maximum = StockConfig.maxDividendPercentFromPrice.coerceAtMost(ABSOLUTE_MAX_DIVIDEND_RATE)
        if (!configured.isFinite() || !maximum.isFinite()) return 0.0
        return configured.coerceIn(0.0, maximum.coerceAtLeast(0.0))
    }

    internal fun effectiveDividendPeriodSeconds(): Long =
        StockConfig.dividendPeriod.coerceAtLeast(MINIMUM_DIVIDEND_PERIOD_SECONDS)

    internal fun effectiveDividendPerShare(stock: Stock): Double {
        val dividend = stock.price * effectiveDividendRate()
        return if (dividend.isFinite() && dividend > 0.0) dividend else 0.0
    }

    internal fun effectiveMaxLeverage(stock: Stock): Int = stock.maxLeverage.coerceIn(1, MAX_NEW_ORDER_LEVERAGE)

    internal fun effectiveMaxBuyPrice(): Double =
        finitePositiveLimit(StockConfig.maxBuyPrice, MAX_NEW_ORDER_PRINCIPAL)

    internal fun effectiveMaxLeveragedPrice(): Double =
        finitePositiveLimit(StockConfig.maxLeveragedPrice, MAX_NEW_ORDER_EXPOSURE)

    internal fun isDividendDue(stock: Stock, now: Long): Boolean =
        now - stock.lastTimeDividend >= effectiveDividendPeriodSeconds() * 1_000L

    private fun finitePositiveLimit(configured: Double, absolute: Double): Double =
        if (configured.isFinite() && configured > 0.0) configured.coerceAtMost(absolute) else absolute

    fun saveHistory() {
        info("Saving stock history")
        HistoryManager.saveHistory()
    }

    private const val MINIMUM_DIVIDEND_PERIOD_SECONDS = 24L * 60L * 60L
    private const val ABSOLUTE_MAX_DIVIDEND_RATE = 0.0002
    private const val MAX_NEW_ORDER_LEVERAGE = 10
    private const val MAX_NEW_ORDER_PRINCIPAL = 100_000.0
    private const val MAX_NEW_ORDER_EXPOSURE = 1_000_000.0
}
